Grouping sets dev

taylor12805 opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
add grouping sets function, like GROUP BY grouping sets (a, b, (a, b))


Detailed description / Documentation draft:
Fix multiple grouping sets element issue from PR: https://github.com/ClickHouse/ClickHouse/pull/24172

MaxTheHuman committed:
init commit with parsing and BAD realisation

MaxTheHuman committed:
rm unnessesary data committed by mistake

MaxTheHuman committed:
remove another unnessesary files

MaxTheHuman committed:
revert changes made to cube transform

MaxTheHuman committed:
fix typo

MaxTheHuman committed:
erase blank line to restore initial state

MaxTheHuman committed:
fixes

MaxTheHuman committed:
fix typos

MaxTheHuman committed:
feat grouping-sets: initial changes

MaxTheHuman committed:
development

MaxTheHuman committed:
grouping sets development

MaxTheHuman committed:
grouping sets development

MaxTheHuman committed:
grouping sets development

MaxTheHuman committed:
grouping sets cleanup

MaxTheHuman committed:
grouping sets: fix 'Port already connected' error

MaxTheHuman committed:
grouping sets dev: fix errors, something works

MaxTheHuman committed:
grouping sets: make simple aggregation with grouping sets to work

MaxTheHuman committed:
grouping sets: fix

MaxTheHuman committed:
grouping sets: add tests, fix bug

MaxTheHuman committed:
grouping-sets: rearrange result columns so that resize is possible after grouping sets

MaxTheHuman committed:
make grouping sets work with total

Commit:
grouping sets support multiple group

taylor12805 commented:
@KochetovNicolai Hi, please take a look at my implementation if you are free

taylor12805 commented:
I find this implementation has performance issue because in AggregatingStep pipeline is resize to 1 and fork to number of grouping sets, which drastically decrease the performance

Commit:
grouing sets transformer instead of fork processor in aggregating transformer

taylor12805 commented:
@KochetovNicolai  Hi, I come up with another implementation in Query Execution phase, using grouping sets transformer. The performance of this implementation is much higher than the one in https://github.com/ClickHouse/ClickHouse/pull/24172, please take a look to check if it is ok.

Commit:
fix bug when no data in grouping sets transformer

Commit:
correct test case results

Commit:
some changes

Commit:
refactor

taylor12805 commented on src/Interpreters/InterpreterSelectQuery.cpp:
```diff
@@ -2156,36 +2229,93 @@ void InterpreterSelectQuery::executeTotalsAndHaving(
+    ColumnNumbers all_keys;
+    ColumnNumbersTwoDimension keys_vector;
+    auto & query = getSelectQuery();
+    if (query.group_by_with_grouping_sets)
```
These two fragment of code are not entirely the same since there are some parameter difference. I'm not sure if we need to move it to a separate function

novikd replied:
I do't see any problem to make these parameters function arguments.

taylor12805 commented:
There is bug when query on Distributed Table with grouping sets function, will try to fix it soon

novikd commented on src/Interpreters/InterpreterSelectQuery.cpp:
```diff
@@ -2069,25 +2073,69 @@ void InterpreterSelectQuery::executeAggregation(QueryPlan & query_plan, const Ac
+    auto & query = getSelectQuery();
+    if (query.group_by_with_grouping_sets)
+    {
+        ColumnNumbers keys;
```
Seems like you don't need this variable anymore.

novikd commented on src/Interpreters/InterpreterSelectQuery.cpp:
```diff
@@ -2169,37 +2216,78 @@ void InterpreterSelectQuery::executeRollupOrCube(QueryPlan & query_plan, Modific
-        keys.push_back(header_before_transform.getPositionByName(key.name));
-
     const Settings & settings = context->getSettingsRef();
+    std::shared_ptr<Aggregator::Params> params_ptr;
```
Use `std::unique_ptr` here because you don't share the ownership.

Commit:
1. refactor ExpressionAnalyzer
2. Update formatAST to support grouping sets query with distributed table
3. modify astExpression to support function in grouping sets

Commit:
update test reference

taylor12805 commented:
@novikd  Hi, I generate reference file 01883_with_grouping_sets.reference following doc: https://clickhouse.tech/docs/en/development/adding_test_queries/, I have no idea why the test fails, please give me a hint

Commit:
update test reference

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=26869) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=26869) before we can accept your contribution.<br/>**1** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: KochetovNicolai<br/>:x: fanzhou<br/>:x: MaxTheHuman<br/><hr/>**fanzhou** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=26869) it.</sub>

novikd committed:
fix GROUPING SETS formating

novikd committed:
remove ya.make

novikd committed:
Cleanup code

novikd committed:
Remove unused code

novikd committed:
Fix removing constant columns

azat commented on src/QueryPipeline/Pipe.h:
```diff
@@ -136,6 +137,8 @@ class Pipe
     /// It is needed for debug. See QueryPipelineProcessorsCollector.
     Processors * collected_processors = nullptr;
 
+    Poco::Logger * log = &Poco::Logger::get("Pipe");
```
Looks unused.

azat commented on src/Processors/QueryPlan/AggregatingStep.h:
```diff
@@ -56,6 +56,8 @@ class AggregatingStep : public ITransformingStep
 
     Processors aggregating;
 
+    Poco::Logger * log = &Poco::Logger::get("AggregatingStep");
```
This one also unused.

azat commented on src/Interpreters/ExpressionAnalyzer.h:
```diff
@@ -94,6 +95,8 @@ class ExpressionAnalyzer : protected ExpressionAnalyzerData, private boost::nonc
         explicit ExtractedSettings(const Settings & settings_);
     };
 
+    Poco::Logger * poco_log = &Poco::Logger::get("ExpressionAnalyzer");
```
Unused

azat commented on src/Processors/Transforms/GroupingSetsTransform.h:
```diff
@@ -0,0 +1,35 @@
+
+    UInt64 keys_vector_idx = 0;
+
+    Poco::Logger * log = &Poco::Logger::get("GroupingSetsTransform");
```
And this one is also unused.

sevirov commented:
Internal documentation ticket: DOCSUP-20353

azat commented on src/Parsers/ASTSelectQuery.cpp:
```diff
@@ -125,6 +128,18 @@ void ASTSelectQuery::formatImpl(const FormatSettings & s, FormatState & state, F
     if (group_by_with_cube)
         s.ostr << (s.hilite ? hilite_keyword : "") << s.nl_or_ws << indent_str << (s.one_line ? "" : "    ") << "WITH CUBE" << (s.hilite ? hilite_none : "");
 
+    if (group_by_with_grouping_sets)
+    {
+        frame.surround_each_list_element_with_parens = true;
+        s.ostr << (s.hilite ? hilite_keyword : "") << s.nl_or_ws << indent_str << (s.one_line ? "" : "    ") << "GROUPING SETS" << (s.hilite ? hilite_none : "");
+        s.ostr << " (";
+        s.one_line
+        ? groupBy()->formatImpl(s, state, frame)
+        : groupBy()->as<ASTExpressionList &>().formatImplMultiline(s, state, frame);
+        s.ostr << ")";
+        frame.surround_each_list_element_with_parens = false;
+    }
```
Formatting looks odd:
```sql
SELECT
    key1,
    key2,
    sum(value)
FROM data
GROUP BY
    GROUPING SETS (
    ( key1),
    ( key2))
```

So looks like:
- `surround_each_list_element_with_parens` should not be set
- extra parenthesis is not required

So something like this:

```c
if (group_by_with_grouping_sets)
{
    s.ostr << (s.hilite ? hilite_keyword : "") << s.nl_or_ws << indent_str << (s.one_line ? "" : "    ") << "GROUPING SETS" << (s.hilite ? hilite_none : "");
    s.one_line
    ? groupBy()->formatImpl(s, state, frame)
    : groupBy()->as<ASTExpressionList &>().formatImplMultiline(s, state, frame);
}
```

novikd replied:
> extra parenthesis is not required

Why? This is part of GROUPING SETS syntax.

But I agree I should remove new line before GROUPING SETS and redundant whitespace before the first element in grouping set.

azat replied:
>Why? This is part of GROUPING SETS syntax.

I don't see why they are required, both queries works fine:
- `SELECT key1, key2, sum(value) FROM data GROUP BY GROUPING SETS (( key1), ( key2))`
- `SELECT key1, key2, sum(value) FROM data GROUP BY GROUPING SETS key1, key2`

Plus `GROUP BY` keys are not surrounded with extra parenthesis.

novikd replied:
Because grouping sets support several grouping commands. 
This is a valid syntax:
```
GROUPING SETS ((key1), (key2), (key1, key2), ())
```
It should work similar to Postgres.

azat replied:
Got it, thanks for the clarifications.
